client/executor/wasmtime/src/runtime.rs

```diff
@@ -158,42 +218,173 @@ directory = \"{cache_dir}\"
+	config
+}
+
+pub struct Semantics {
+	/// Enabling this will lead to some optimization shenanigans that make calling [`WasmInstance`]
+	/// extermely fast.
+	///
+	/// Primarily this is achieved by not recreating the instance for each call and performing a
+	/// bare minimum clean up: reapplying the data segments and restoring the values for global
+	/// variables. The vast majority of the linear memory is not restored, meaning that effects
+	/// of previous executions on the same [`WasmInstance`] can be observed there.
+	///
+	/// This is not a problem for a standard substrate runtime execution because it's up to the
+	/// runtime itself to make sure that it doesn't involve any non-determinism.
+	///
+	/// Since this feature depends on instrumentation, it can be set only if [`CodeSupplyMode::Verbatim`]
+	/// is used.
+	pub fast_instance_reuse: bool,
+
+	/// The WebAssembly standard defines a call/value stack but it doesn't say anything about its
+	/// size except that it has to be finite. The implementations are free to choose their own notion
+	/// of limit: some may count the number of calls or values, others would rely on the host machine
+	/// stack and trap on reaching a guard page.
+	///
+	/// This obviously is a source of non-determinism during execution. This feature can be used
+	/// to instrument the code so that it will count the depth of execution in some deterministic
+	/// way (the machine stack limit should be so high that the deterministic limit always triggers
+	/// first).
+	///
+	/// See [here][stack_height] for more details of the instrumentation
+	///
+	/// Since this feature depends on instrumentation, it can be set only if [`CodeSupplyMode::Verbatim`]
+	/// is used.
+	///
+	/// [stack_height]: https://github.com/paritytech/wasm-utils/blob/d9432baf/src/stack_height/mod.rs#L1-L50
+	pub stack_depth_metering: bool,
+	// Other things like nan canonicalization can be added here.
+}
+
+pub struct Config {
+	/// The number of wasm pages to be mounted after instantiation.
+	pub heap_pages: u32,
+
+	/// The WebAssembly standard requires all imports of an instantiated module to be resolved,
+	/// othewise, the instantiation fails. If this option is set to `true`, then this behavior is
+	/// overriden and imports that are requested by the module and not provided by the host functions
+	/// will be resolved using stubs. These stubs will trap upon a call.
+	pub allow_missing_func_imports: bool,
+
+	/// A directory in which wasmtime can store its compiled artifacts cache.
+	pub cache_path: Option<PathBuf>,
+
+	/// Tuning of various semantics of the wasmtime executor.
+	pub semantics: Semantics,
+}
+
+pub enum CodeSupplyMode<'a> {
+	/// The runtime is instantiated using the given runtime blob.
+	Verbatim {
+		// Rationale to take the `RuntimeBlob` here is so that the client will be able to reuse
+		// the blob e.g. if they did a prevalidation. If they didn't they can `RuntimeBlob`
```

Comment:
> If they didn't they can `RuntimeBlob`

Needs to be fixed.

In general I'm a little bit confused by this whole docs in here.

Reply:
What is confusing? Do you have a suggestions how to improve that?

So what I am trying to say there, is that there are two use cases how sc-executor-wasmtime is going to be invoked: the traditional substrate path that uses this fast instance reuse thing and PVF execution. Both of those require to do some instrumentation under the hood, meaning that `RuntimeBlob` will be required.

However, I cannot say that this will be the case forever, at some point we may not need to do instrumentation nor prevalidation in which case creation of the runtime blob will be superfluous. In that case I suggest in the comment we can introduce a `Cow` like thing, which will create `RuntimeBlob` only if it's not provided and if it is required.